React/DP-15689:  Extend ButtonWithIcon & Add LinkDropdown

avrilpearl opened the pull request:
Minor

Changed
- (React) [ButtonWithIcon] DP-15689: Extended the ButtonWithIcon component to pass multiple theme (c-primary, c-primary-alt, c-highlight, c-gray-dark) and display for multiple usages per base button atom (primary, secondary, tertiary, quaternary, quaternary-simple). #730
- (React) [DecorativeLink] DP-15689: Enabled classnames to be passed to DecorativeLink. #730
- (React) [ListingTable] DP-15689: Updated ListingTable to consume ButtonWithIcon for expand functionality. #730
- (React) [ButtonAlert] DP-15689: Made buttonalert a variant of buttonwithicon. #730

Added
- (React) [LinkDropdown] DP-15689: Created LinkDropdown molecule. #730

Fixed
- (React) [Button, ButtonWithIcon] DP-15689: Fixed color contrast a11y issue of c-highlight theme. #730

isaacchansky commented on styleguide/source/assets/scss/06-theme/03-organisms/_about-section.scss:
```diff
@@ -0,0 +1,103 @@
+
+    .ma__content-link  {
+      display: inline-block;
+      margin: 5px 0 2.5rem;
```
do you think we can use rem instead of pixels here?

legostud replied:
rem seems a bit off.  I could see ems if the spacing needs to change based on the items font-size.

legostud commented on styleguide/source/_patterns/03-organisms/by-author/person-card.twig:
```diff
@@ -0,0 +1,23 @@
+<div class="ma__person-card">
+  <div class="ma__person-card--wrapper">
```
In correct BEM syntax => `ma__person-card--wrapper` would be a modifier not a block element

I'm guessing you wanted ma__person-card`__wrapper`.  Otherwise the modifier should be added to the element with the same base class `class="ma__person-card ma__person-card--wrapper`

legostud commented on styleguide/source/_patterns/03-organisms/by-author/person-card.twig:
```diff
@@ -0,0 +1,23 @@
+        {% set compHeading = person.name %}
+        {% include "@atoms/04-headings/comp-heading.twig" %}
+      </div>
+      <div class="ma__person-card--title">
```
should be `__title`

legostud commented on styleguide/source/assets/scss/03-organisms/_person-card.scss:
```diff
@@ -0,0 +1,77 @@
+
+  &__person-info {
+
+    .ma__person-card--name {
```
Same issue as `--wrapper`

legostud commented on styleguide/source/assets/scss/03-organisms/_person-card.scss:
```diff
@@ -0,0 +1,77 @@
+      }
+    }
+
+    .ma__person-card--title {
```
same issue as `--wrapper`

legostud commented on styleguide/source/assets/scss/03-organisms/_person-card.scss:
```diff
@@ -0,0 +1,77 @@
+      margin: 0 auto;
+    }
+  }
+}
```
missing line ending - add an extra line-break.

legostud commented on styleguide/source/assets/scss/06-theme/03-organisms/_about-section.scss:
```diff
@@ -0,0 +1,103 @@
+      background-color: $c-bg-section;
+    }
+
+    .ma__about-section__content-container {
```
I think this would just be `&__content-container` outside of the `&__content` selector

legostud commented on styleguide/source/assets/scss/06-theme/03-organisms/_about-section.scss:
```diff
@@ -0,0 +1,103 @@
+      }
+    }
+
+    .ma__about-section__content-text {
```
`&__content-text` outside of `&__content`  unless there is a reason why you need it to be `.ma__about-section__content .ma__about-section__content-text` at which point you can do `&__content &__content-text`

legostud commented on styleguide/source/assets/scss/06-theme/03-organisms/_about-section.scss:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  .ma__person-card {
```
placement seems a bit random to me.  I'm worried this override will get lost among the other selectors.  Not sure if there is a better placement.

clair0917 replied:
My thought process was that this placement is about the layout of the about section, not the person card. Since these styles have no affect on the person card, I thought this was the most logic place to put it.

legostud replied:
makes sense.

legostud commented on styleguide/source/assets/scss/03-organisms/_person-card.scss:
```diff
@@ -18,7 +18,7 @@
   &__person-info {
 
-    .ma__person-card--name {
+    .ma__person-card__name {
```
I don't understand why this needs to be nested within `&__person-info`.  I think it could be at the root level `&__name` instead.

legostud commented on styleguide/source/assets/scss/03-organisms/_person-card.scss:
```diff
@@ -35,7 +35,7 @@
     }
 
-    .ma__person-card--title {
+    .ma__person-card__title {
```
I don't understand why this needs to be nested within `&__person-info`.  I think it could be at the root level `&__title` instead.

avrilpearl committed:
react/DP-15689: Extend decorativelink to pass classname

avrilpearl committed:
react/DP-15689--extend-buttonwithicon

avrilpearl committed:
react/DP-15689--add-linkdropdown-molecule

avrilpearl committed:
react/DP-15689--add-linkdropdown-style

avrilpearl committed:
react/DP-15689: Add changelog

avrilpearl committed:
add aria-expanded

avrilpearl committed:
update listing table molecule

avrilpearl committed:
Update changelog

avrilpearl committed:
react/DP-15689--fix-c-highlight-a11y

avrilpearl committed:
fix typeahead dropdown component

avrilpearl committed:
react/DP-15689--buttonalert-variant-buttonwithicon

avrilpearl committed:
update screenshots

avrilpearl committed:
fix lint errors

avrilpearl committed:
update changelog and add to export

avrilpearl committed:
update padding

avrilpearl committed:
update svg transparency

avrilpearl committed:
simpify scss

avrilpearl committed:
update style

clairesunstudio commented on assets/scss/00-base/mixins/_ma-button.scss:
```diff
@@ -33,7 +33,13 @@
       border-color: transparent;
-      color: $c-font-inverse;
+
+      @if $color == $c-highlight {
```
thanks for fixing this! I just realized that the secondary highlight button is also not accessible. but I don't think it looks good with black text either.
<img width="148" alt="Screen Shot 2019-09-06 at 3 33 06 PM" src="https://user-images.githubusercontent.com/5789411/64455493-cd19e980-d0bb-11e9-873e-989968284c73.png">
<img width="144" alt="Screen Shot 2019-09-06 at 3 31 17 PM" src="https://user-images.githubusercontent.com/5789411/64455689-4d404f00-d0bc-11e9-96f8-15c610f7759d.png">

we could add a darker yellow color into the design system. e.g.
<img width="151" alt="Screen Shot 2019-09-06 at 3 36 22 PM" src="https://user-images.githubusercontent.com/5789411/64455737-6943f080-d0bc-11e9-9461-d6b386f3adcf.png">

my recommendation would be to remove that option from secondary. WIll leave up to you and @ygannett to decide

avrilpearl replied:
updated to black - will wait to hear from yukiko to finish this

ygannett replied:
I think it's better to remove the option.  Yellow is a tricky color. I looked through the other options.  Their pattern is dark text on a lighter version of the theme colors at normal state, then white text on the theme colors. Even with a darker yellow for the text, we cannot follow this pattern. It would look off from the pattern.

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -45,24 +89,149 @@
     vertical-align: middle;
     fill: $c-gray-dark;
```
```suggestion
```
currently this class is interfering iconColor setting

avrilpearl replied:
?

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -1,10 +1,54 @@
+  "c-gray-dark": $c-gray-dark,
+  "c-white": $c-white
+);
+@mixin ma-button-icon-secondary ($color: $c-primary-alt) {
```
can you move these @mixins into mixins folder? just so it stays consistent with _buttons.scss

maybe see if these make sense to be somehow combined in the _ma-button mixin (just a thought)

avrilpearl replied:
removed mixins here.

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -1,10 +1,54 @@
+    border-bottom-color: $color;
+  }
+}
+@mixin ma-button-icon-quaternary-simple ($color: $c-primary-alt) {
```
In regular button component, quaternary is hard coded to only render theme `$c-primary` (all the other theme colors don't work) 
I think we should extend the buttons to allow button quarternary to use more themes except for hightlight, as currently they don't work in all quarternary usage for accessiblity
<img width="136" alt="Screen Shot 2019-09-06 at 3 55 15 PM" src="https://user-images.githubusercontent.com/5789411/64457236-12d8b100-d0c0-11e9-81cc-bb63743d94c8.png">
<img width="146" alt="Screen Shot 2019-09-06 at 3 55 18 PM" src="https://user-images.githubusercontent.com/5789411/64457237-13714780-d0c0-11e9-80a0-13bfa61c7033.png">

avrilpearl replied:
removed option for quaternary to use c-highlight theme

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -45,24 +89,149 @@
 
-  &.ma__icon-green > svg {
-    fill: $c-primary-alt;
+  @each $class, $c in $colors {
```
i'm wondering if setting the icon svg color separately, as I feel like the icon should be aligned with the color of the button text and set by the theme of the button in the mixin as a whole (including hover state). I can't think of a case it's should be different on the top of my head. e.g. tertiary button if themes can't be changed, icon color shouldn't be able to change either

avrilpearl replied:
removed option to style icon with icon color prop

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -45,24 +89,149 @@
+    border: none;
+    padding: 0;
+    text-transform: none;
+    border-bottom: 2px solid rgba($c-primary-alt,0.75);
```
setting the default here is not neccessary, as in the mixin you already set default by passing `$color: $c-primary-alt`

clairesunstudio replied:
also applied to all usage, don't need to set default color as its already set in mixin

avrilpearl replied:
@clairesunstudio --> not true here since default theme is c-primary-alt that theme class is not appended so I need to set the default with no theme class to c-primary-alt. Unless we want to change this but it is the same way the base button is doing it (buttons.scss), where with no class it is styled in the theme c-primary.

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -1,10 +1,54 @@
+    border-bottom-color: $color;
+  }
+}
+@mixin ma-button-icon-quaternary-simple ($color: $c-primary-alt) {
```
can we set the default to $c-primary instead
```suggestion
@mixin ma-button-icon-quaternary-simple ($color: $c-primary) {
```

avrilpearl replied:
Please see the response to mixin comment above, do to this I cannot change the default theme away from primary-alt

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -1,10 +1,54 @@
+    fill: rgba($color,0.5);
+  }
+}
+@mixin ma-button-icon-alert ($color: $c-primary-alt) {
```
can we set the default to $c-black instead
```suggestion
@mixin ma-button-icon-alert ($color: $c-black) {
```

avrilpearl replied:
see the response to mixin comment above re: default color.

clairesunstudio commented on react/src/components/atoms/buttons/ButtonWithIcon/style.scss:
```diff
@@ -1,10 +1,54 @@
+    }
+  }
+}
+@mixin ma-button-icon-quaternary ($color: $c-primary-alt) {
```
can we set the default to $c-primary instead
```suggestion
@mixin ma-button-icon-quaternary ($color: $c-primary) {
```

avrilpearl replied:
see the response to mixin comment above re: default color.

clairesunstudio commented on react/src/components/molecules/LinkDropdown/index.js:
```diff
@@ -0,0 +1,118 @@
+      this.props.onItemSelect({ target });
+    }
+  }
+  handleClickOutside(event) {
```
should escape key also close dropdown?

avrilpearl replied:
added this.

clairesunstudio commented on react/src/components/molecules/LinkDropdown/index.js:
```diff
@@ -0,0 +1,118 @@
+  render() {
+    const { dropdownItems, dropdownButton } = this.props;
+    const dropdownButtonProps = {
+      ...dropdownButton,
```
should `'aria-haspopup': ` in this component also be true, without the option to overwrite?

avrilpearl replied:
always set to true

clairesunstudio commented on react/src/components/molecules/LinkDropdown/LinkDropdown.stories.js:
```diff
@@ -0,0 +1,33 @@
+        theme: select('theme', buttonWithIconOptions.theme, 'c-primary', 'dropdownButton'),
+        usage: select('usage', buttonWithIconOptions.usage, 'quaternary-simple', 'dropdownButton'),
+        id: text('id', 'dropdownbutton-simple', 'dropdownButton'),
+        'aria-haspopup': boolean('aria-haspopup', true, 'dropdownButton'),
```
shouldn't this always be true?

clairesunstudio commented on react/src/components/molecules/LinkDropdown/index.js:
```diff
@@ -0,0 +1,118 @@
+        <ButtonWithIcon {...dropdownButtonProps} />
+        {this.state.buttonExpand && (
+          <div className={dropdownMenuClasses} aria-labelledby={dropdownButtonProps.id}>
+            {dropdownItems.map((item) => <DecorativeLink {...item} className={dropdownItemClasses} onClick={this.handleSelect} />)}
```
this handleSelect function will never get triggered because onClick is not a prop for DecorativeLink. If we want to pass a custom click function here we should probably use Button

clairesunstudio commented:
linkDropDown hover style is always primary alt, even a different theme is set. the text color on hover should probably match the theme selected
<img width="153" alt="Screen Shot 2019-09-06 at 5 13 00 PM" src="https://user-images.githubusercontent.com/5789411/64460991-e7f35a80-d0c9-11e9-9e8f-d12f80b45c23.png">

should we move buttonWithIcon styles into assets folder?

clairesunstudio commented:
Great work! Left quite a few comments above don't hate me 😛You refactored a bunch of things 👍  !

avrilpearl committed:
Merge branch 'develop' into react/DP-15689--extend-buttonwithicon-add-dropdown

avrilpearl committed:
add linkdropdown review comments

[LinkDropdown]
- escape key also close dropdown
- 'aria-haspopup' always true
- click function here we should probably use Button

avrilpearl committed:
fix lint warning

avrilpearl committed:
address button with icon scss notes

avrilpearl committed:
rename scss file

avrilpearl committed:
fix base button with icon primary styling

ygannett reviewed (commented):
Looks good!
Please see my comment for assets/scss/00-base/mixins/_ma-button.scss.

avrilpearl committed:
Merge branch 'develop' into react/DP-15689--extend-buttonwithicon-add-dropdown

avrilpearl committed:
address yukiko's comment remove c-highlight on secondary button

ygannett reviewed (approved):
Looks good.
